Система постов

NadirCianid opened the pull request.

NadirCianid committed:
Реализовал требуемые действия без тестов

NadirCianid committed:
Реализовал требуемые действия без тестов

AlexHor230 commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,203 @@
+        checkUserExistence(userId);
+
+        return postRepository.findByAuthorId(userId).stream()
+                .filter(post -> post.getAuthorId().equals(userId))
```
Я думаю фильтр лишний, учитывая что ты уже проверяешь, что пользователь существует - checkUserExistence, после чего фильтруешь findByAuthorId(userId)

AlexHor230 commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,203 @@
+        checkProjectExistence(projectId);
+
+        return postRepository.findByProjectId(projectId).stream()
+                .filter(post -> post.getProjectId().equals(projectId))
```
См. коммент выше

NadirCianid committed:
Исправил замечания:
- убрал лишние фильтрации в методах сервиса get...

Sabina-Vishnya commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,89 @@
+    public PostDto updatePost(@Valid @RequestBody PostDto postDto) {
+        checkAuthorPresence(postDto);
+
+        if (postDto.getId() == null) {
```
почему не аннотацией?

NadirCianid replied:
Потому что это поле не всегда должно быть NonNull. Только при обновлении

Sabina-Vishnya commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,89 @@
+        return postService.getPostsOfProject(projectId);
+    }
+
+    private void checkAuthorPresence(PostDto postDto) {
```
тот же вопрос

NadirCianid replied:
здесь поля могут быть null, но только один любой из двух

Sabina-Vishnya commented on src/main/java/faang/school/postservice/controller/post/PostController.java:
```diff
@@ -0,0 +1,89 @@
+@RestController
+@RequestMapping("/post")
+@RequiredArgsConstructor
+public class PostController {
```
без ResponseEntity всяких почему-то
и @ExceptionHandler - про это уроки есть на платформе

Sabina-Vishnya commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,203 @@
+@Slf4j
+public class PostService {
+    private final PostRepository postRepository;
+    private final LikeRepository likeRepository;
```
зачем тут два неиспользуемых репозитория?

Sabina-Vishnya commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,203 @@
+                .filter(Post::isPublished)
+                .sorted(Comparator.comparing(Post::getPublishedAt).reversed())
+                .map(postMapper::toDto)
+                .toList();
```
в нескольких местах дублирование, избавляемся от этого

Sabina-Vishnya commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    private void checkPostMatchingWithSystem(PostDto postDto, Post postToBeUpdated) {
```
что-то можно утащить в аннотации, даже условно сервисного слоя, чтобы разделять так
но не писать только именно Post энтити в параметрах вместе с дто, а брать именно поле (выходит параметр второй можеет быть лонг какой-нибудь, а первых дто)

про аннотации ты знаешь, давай кастомные начнем делать
только проверяй работоспособность и не забудь через постман покидать запросов, чтобы убедиться, что аннотация кастомная и любая другая, работает и отстреливает ошибкой, в случае кривых данных

NadirCianid replied:
я пока не понимаю, как эту логику можно реализовать через аннотации, если аннотации задаются на уровне класса, где у меня не будет доступа к тем данным с которыми надо сверяться

NadirCianid committed:
Исправил замечания:
- добавил глобальный обработчик исключений
- убрал дублирование в методах получения постов

NadirCianid committed:
Исправил замечания:
- добавил глобальный обработчик исключений
- убрал дублирование в методах получения постов

Sabina-Vishnya reviewed (approved):
и тесты

Sabina-Vishnya commented:
верну, чтобы не болталось в ревью, а когда вольешь в done уже перекинешь

NadirCianid committed:
Изменил названия исключений на более общие и перенес енамы с сообщениями об ошибках в отдельный подпакет

NadirCianid committed:
Изменил названия исключений на более общие и перенес енамы с сообщениями об ошибках в отдельный подпакет

NadirCianid committed:
Merge branch 'phoenix-master-bc4' into feature-BJS2-6537

# Conflicts:
#	src/main/java/faang/school/postservice/dto/post/PostDto.java
#	src/main/java/faang/school/postservice/exception/DataValidationException.java
#	src/main/java/faang/school/postservice/exception/GlobalExceptionHandler.java
#	src/main/java/faang/school/postservice/mapper/PostMapper.java

NadirCianid committed:
Написал тесты

Sabina-Vishnya commented on src/main/java/faang/school/postservice/service/post/PostVerifier.java:
```diff
@@ -0,0 +1,87 @@
+
+@Component
+@RequiredArgsConstructor
+class PostVerifier {
```
Ну вот коммит сообщает о том, что ты тесты написал, а по факту, еще и рефакторинг сделал)
почему бы не указать это в коммите или отдельной таской тогда не сделать? 

И метод verifyPostMatchingWithSystem подрефачить бы, а то длинный, какие-то проверки обобщить и вынести в приватный метод с параметрами.
Еще можно подумать над тем, чтобы что-то вынести в метод, но в модель уже унести, из каких-то проверок.  Немного в ООП стиле сделать, тогда похудеет класс, но это если получится, тут в целом можно это сделать, а так хотя бы просто метод подрефачить

Sabina-Vishnya commented on src/test/java/faang/school/postservice/service/post/PostServiceTest.java:
```diff
@@ -0,0 +1,278 @@
+        draftsOfProject = testData.getDraftsOfProject();
+        postsOfUser = testData.getPostsOfUser();
+        postsOfProject = testData.getPostsOfProject();
+
```
отступ лишний

Sabina-Vishnya commented on src/test/java/faang/school/postservice/util/TestData.java:
```diff
@@ -35,4 +38,60 @@ public List<Comment> returnListOfComments() {
+                .build();
+    }
+
+    public List<Post> getDraftsOfUser() {
```
подобные методы вызывают метод returnPost, который билд содержит, дальше делают билд уже без метода, потому что другие значения нужны объекту - это излишни, тогда в чем смысл метода returnPost?
можно один метод в параметрами сделать и тогда ты будешь вызывать условный createPost(var 1, var 2, var 3) с какими-то параметрами, куда будешь уже нужное значение класть

также всякие LocalDateTime повторяются, это может быть константами и вызываться там, где нужно
под константы можно завести отдельный класс, чтобы разделять, а можно пока в этот пихать, но тогда понамшено будет и со временем сильнее разрастется

NadirCianid committed:
Исправил ошибку после rebase

NadirCianid committed:
Merge branch 'phoenix-master-bc4' into feature-BJS2-6537

NadirCianid committed:
Исправил ошибку после rebase

NadirCianid committed:
Исправил замечания по тестам

Sabina-Vishnya commented on src/test/java/faang/school/postservice/util/TestData.java:
```diff
@@ -11,28 +12,86 @@
+                .build();
+    }
+
+    public List<Post> getDraftsOfUser() {
```
награмождать тестовый класс тоже не стоит прям, этот метод и ниже с List.of можно было не делать, а просто List.of и твой returnPostCreatedByUser использовать в тестовом классе, но это так, можно не править

Sabina-Vishnya commented on src/main/java/faang/school/postservice/service/post/PostVerifier.java:
```diff
@@ -0,0 +1,102 @@
+
+@Component
+@RequiredArgsConstructor
+class PostVerifier {
```
эх, название Verifier во всех валидаторах всё таки смущает меня и не очень о валидаторе говорит будто, но это вкусовщина скорее, вливали уже с Verifier, так что норм

Sabina-Vishnya reviewed (approved):
нормас, вливай
